MESOS: Send graceful termination update from executor to kubelet

sttts opened the pull request:
The kubelet gets pod updates from the executor. The executor only passed ADD and REMOVE updates, it did not notice a graceful termination. Hence, terminating pods stayed terminating forever.

With this PR the executor watches the pods which have a binding to the node. When a pod is updated with `pod.DeletionTimestamp != nil` the executor compares it with the pod object it sent to the kubelet. If this differs, the kubelet is notified with a pod update for this node.

Fixes https://github.com/mesosphere/kubernetes-mesos/issues/495

sttts commented:
/cc @s-urbaniak @jdef for internal review

jdef commented on contrib/mesos/pkg/scheduler/plugin.go:
```diff
@@ -578,11 +598,52 @@ func (k *deleter) Run(updates <-chan queue.Entry, done <-chan struct{}) {
 				k.qr.updatesAvailable()
+
+				if entry.Is(queue.UPDATE_EVENT) && pod.InGracefulTermination() {
+					if err := k.terminateOne(pod); err != nil {
```
why terminateOne instead of KillTask?

jdef commented on contrib/mesos/pkg/scheduler/plugin.go:
```diff
@@ -103,6 +106,23 @@ func (k *k8smScheduler) killTask(taskId string) error {
+	}
+
+	data := fmt.Sprintf("%s:%s", messages.TaskTerminate, string(msg)) //TODO(jdef) use a real message type
+	_, err = k.internal.driver.SendFrameworkMessage(task.Executor.GetExecutorId(), mutil.NewSlaveID(task.Spec.SlaveID), data)
```
normally we'd invoke KillTask on the driver .. why don't we do that here?

sttts replied:
it's termination in the sense of graceful termination, i.e. SIGTERM, not SIGKILL.

sttts replied:
In the Mesos sense, the task keep running.

jdef replied:
nevermind, i see that we're passing data through to the executor so that it can tell the kubelet about graceful termination.

framework messages aren't reliable. while this implementation is probably fine for now as a first step, we'll want to file a ticket to add reliability guarantees around the delivery of messages like this.

sttts replied:
An alternative would be to watch the pods for a node in the executor and notify the kubelet about them. This is essentially what the kubelet would do if we used an apiserver config source. I think I would prefer that over this solution in the scheduler.

jdef replied:
:+1: for the executor-watch suggestion. closer to the native k8s arch and less flaky in terms of messaging.

sttts replied:
and less complexity because it's completely local to the executor.

jdef commented on contrib/mesos/pkg/scheduler/podtask/pod_task.go:
```diff
@@ -59,10 +59,10 @@ type T struct {
 	CreateTime  time.Time
-	UpdatedTime time.Time // time of the most recent StatusUpdate we've seen from the mesos master
+	UpdatedTime time.Time           // time of the most recent StatusUpdate we've seen from the mesos master
+	Executor    *mesos.ExecutorInfo // readonly
```
dislike making this public. encourages callers to mess with its state, which I expressly do NOT want to encourage. would rather see an `ExecutorID() string` func added to task, since it looks like that's the only bit that's needed outside of this package

sttts replied:
I totally agree. See my previous comment about moving it to the executor.

jdef commented on contrib/mesos/pkg/executor/messages/messages.go:
```diff
@@ -26,9 +30,17 @@ const (
 	LaunchTaskFailed         = "launch-task-failed"
 	TaskKilled               = "task-killed"
+	TaskLost                 = "task-lost"
+	TaskTerminate            = "task-terminate"
```
comment here? perhaps "supports graceful termination"

sttts committed:
Dequeue pods in scheduler which are terminating

sttts committed:
Watch pods in executor and tell kubelet about graceful terminations

- instantiate framework.Controller for pods in the executor using framework.NewInformer,
  in order to watch pod updates for pods on that host
- forwards updates like graceful termination to the kubelet.

  This might also be the preparation for other updates which are supported by the
  kubelet.

sttts commented:
Rebased.

sttts commented:
/cc @jdef ptal

still fighting with shippable....

jdef commented on contrib/mesos/pkg/scheduler/plugin.go:
```diff
@@ -437,7 +437,7 @@ func (q *queuer) Run(done <-chan struct{}) {
 
 			pod := p.(*Pod)
-			if recoverAssignedSlave(pod.Pod) != "" {
+			if recoverAssignedSlave(pod.Pod) != "" || pod.InGracefulTermination() {
```
remind me why we need this here.. how is it possible for a pod to be in graceful termination w/o having been assigned a binding annotation?

sttts replied:
Maybe we can simplify this one. The idea is that a pod is deleted before being scheduled (e.g. because no offer matched). I wonder who will delete such a pod though. Normally the kubelet has the logic to actually delete a pod after termination.

jdef replied:
If it's deleted before it's scheduled can it possibly enter into graceful
termination?

On Mon, Sep 21, 2015 at 1:24 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/scheduler/plugin.go
> <https://github.com/kubernetes/kubernetes/pull/13966#discussion_r39998547>
> :
>
> > @@ -437,7 +437,7 @@ func (q *queuer) Run(done <-chan struct{}) {
> >  			}
> >
> >  			pod := p.(*Pod)
> > -			if recoverAssignedSlave(pod.Pod) != "" {
> > +			if recoverAssignedSlave(pod.Pod) != "" || pod.InGracefulTermination() {
>
> Maybe we can simplify this one. The idea is that a pod is deleted before
> being scheduled (e.g. because no offer matched). I wonder who will delete
> such a pod though. Normally the kubelet has the logic to actually delete a
> pod after termination.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13966/files#r39998547>.
>

sttts replied:
In theory it can if the apiserver does not forbid this. Will check.

sttts replied:
The apiserver checks the NodeName. If it is empty, the deletion is immediate. If it is set, the graceful timeout will apply. In the case the NodeName is set to a non-existing node, the pod will countdown its graceful deletion timeout even in upstream.

We have the special case that we schedule pods with NodeName != "". Even though this case only happens for pre-scheduled pods which are gracefully deleted, I would leave the above predicate. I will add a comment.

jdef replied:
part of being `InGracefulTermination` is that the pod must be in
`PodRunning` phase. Is this phase possible if we're unable to recover the
assigned slave from the pod? If it's not, then how does it serve us to
check for graceful termination status here?

On Tue, Sep 22, 2015 at 9:37 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/scheduler/plugin.go
> <https://github.com/kubernetes/kubernetes/pull/13966#discussion_r40086121>
> :
>
> > @@ -437,7 +437,7 @@ func (q *queuer) Run(done <-chan struct{}) {
> >  			}
> >
> >  			pod := p.(*Pod)
> > -			if recoverAssignedSlave(pod.Pod) != "" {
> > +			if recoverAssignedSlave(pod.Pod) != "" || pod.InGracefulTermination() {
>
> The apiserver checks the NodeName. If it is empty, the deletion is
> immediate. If it is set, the graceful timeout will apply. In the case the
> NodeName is set to a non-existing node, the pod will countdown its graceful
> deletion timeout even in upstream.
>
> We have the special case that we schedule pods with NodeName != "". Even
> though this case only happens for pre-scheduled pods which are gracefully
> deleted, I would leave the above predicate. I will add a comment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13966/files#r40086121>.
>

sttts replied:
I agree. The "running" check must go for my reasoning above.

Moreover, I agree that we are really talking about some corner case:
- pre-scheduled
- gracefully deleted before being scheduler.

In the worst case the pod is 
- either never launched because the node does not exist. Then it stays in the scheduling queue with some backoff between each scheduling run, until the grace period is over
- or it is launched and directly terminated by the kubelet.

What's your suggestion about this? Drop it or keep it?

jdef replied:
better w/ "running" check removed. this is ready for merge

On Tue, Sep 22, 2015 at 10:47 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/scheduler/plugin.go
> <https://github.com/kubernetes/kubernetes/pull/13966#discussion_r40094890>
> :
>
> > @@ -437,7 +437,7 @@ func (q *queuer) Run(done <-chan struct{}) {
> >  			}
> >
> >  			pod := p.(*Pod)
> > -			if recoverAssignedSlave(pod.Pod) != "" {
> > +			if recoverAssignedSlave(pod.Pod) != "" || pod.InGracefulTermination() {
>
> I agree. The "running" check must go for my reasoning above.
>
> Moreover, I agree that we are really talking about some corner case:
>
>    - pre-scheduled
>    - gracefully deleted before being scheduler.
>
> In the worst case the pod is
>
>    - either never launched because the node does not exist. Then it stays
>    in the scheduling queue with some backoff between each scheduling run,
>    until the grace period is over
>    - or it is launched and directly terminated by the kubelet.
>
> What's your suggestion about this? Drop it or keep it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13966/files#r40094890>.
>

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -329,6 +356,43 @@ func (k *KubernetesExecutor) LaunchTask(driver bindings.ExecutorDriver, taskInfo
+func (k *KubernetesExecutor) handleChangedApiserverPod(pod *api.Pod) {
+	taskId := pod.Annotations[meta.TaskIdKey]
+	if taskId == "" {
+		log.V(5).Infof("ignoring pod update for %s/%s because %s annotation is missing", pod.Namespace, pod.Name, meta.TaskIdKey)
```
doesn't this seem like a bug/error condition that we'd want to report on with a warning level?

sttts replied:
A pod can be assigned to this node, but not scheduled ("pre-scheduled pods").

jdef replied:
can we add a comment here along these lines?

On Mon, Sep 21, 2015 at 1:26 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/executor.go
> <https://github.com/kubernetes/kubernetes/pull/13966#discussion_r39998824>
> :
>
> > @@ -329,6 +356,43 @@ func (k *KubernetesExecutor) LaunchTask(driver bindings.ExecutorDriver, taskInfo
> >  	go k.launchTask(driver, taskId, pod)
> >  }
> >
> > +func (k *KubernetesExecutor) handleChangedApiserverPod(pod *api.Pod) {
> > +	taskId := pod.Annotations[meta.TaskIdKey]
> > +	if taskId == "" {
> > +		log.V(5).Infof("ignoring pod update for %s/%s because %s annotation is missing", pod.Namespace, pod.Name, meta.TaskIdKey)
>
> A pod can be assigned to this node, but not scheduled ("pre-scheduled
> pods").
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13966/files#r39998824>.
>

sttts replied:
done

sttts commented:
@jdef Have addressed your two comments.

sttts commented:
Remove the RUNNING check from InGracefulTermination.

jdef commented:
lgtm

karlkfi commented:
This PR fixes several mesos e2e tests that are failing due to pods not shutting down correctly.

gmarek commented:
It's not touching main Kubernetes code and fixes some Mesos tests so it's safe to merge.
